Lib/csharp/std_wstring.i

```diff
@@ -31,8 +31,21 @@ static std::wstring Swig_csharp_UTF16ToWString(const unsigned short *str) {
 
     std::wstring result;
     result.reserve(ptr - pBegin);
-    while(pBegin != ptr)
-      result.push_back(*pBegin++);
+    while (pBegin != ptr) {
+      unsigned short uc = *pBegin++;
+      if (uc <= 0xD7FFu || uc >= 0xE000u) {
+        result.push_back(uc);
+      } else if (pBegin != ptr && (uc & 0xFC00u) == 0xD800u) {
+        unsigned short uc1 = *pBegin++;
+        if ((uc1 & 0xFC00u) == 0xDC00u) {
```

Comment:
In the case of a lone high surrogate pair codepoint followed by a valid character, this will swallow that valid character.

However, Unicode rules are to treat ill-formed data (which unpaired surrogates are) as an error - see https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf C10 page 12 (or 81 by the page numbers in the document).

Simply stripping them like you're attempting to do here can lead to security holes - e.g. a bad string can be smuggled past a vetting check by inserting an unpaired surrogate in the middle of it.

Reply:
I updated the code to insert replacement character 0xFFFD, when wrong surrogate pair were found.

Reply:
@ojwb, please resolve this if you are happy with this commit.

Reply:
Using a replacement character is reasonable for a web browser or similar, but as a general strategy I think following the Unicode rules and flagging this as an error is a better approach (which is what I already suggested above).  

Maybe some users (e.g. who aren't working on data that could be from an attacker or are using the data in a way where there are no security concerns) will want something different, but they can always provide their own typemaps.

(Or maybe we could just delegate the handling of this to the C# runtime, but that only works if it provides an API to do this conversion - I don't know if it does or not.)

Reply:
Most users expect some useful output, replacement characters at least give a level of the secure concerns.
Error handling may be a bit overkill.